src/strawman/collections/CollectionStrawMan4.scala

```diff
@@ -0,0 +1,486 @@
+    def fromIterable[B](c: Iterable[B]): ArrayBuffer[B] = c match {
+      case View.Concat(fst: ArrayBuffer[B], snd: ArrayBuffer[B]) =>
+        val elems = new Array[AnyRef](fst.length + snd.length)
+        Array.copy(fst.elems, fst.start, elems, 0, fst.length)
+        Array.copy(snd.elems, snd.start, elems, fst.length, snd.length)
+        new ArrayBuffer(elems, elems.length)
+      case pd @ View.Partitioned(partition: View.Partition[B]) =>
+        partition.distribute(new ArrayBuffer[B]())
+        pd.forced.get.asInstanceOf[ArrayBuffer[B]]
+      case c if c.knownLength >= 0 =>
+        val elems = new Array[AnyRef](c.knownLength)
+        val it = c.iterator
+        for (i <- 0 until elems.length) elems(i) = it.next().asInstanceOf[AnyRef]
+        new ArrayBuffer[B](elems, elems.length)
+      case _ =>
+        val buf = new ArrayBuffer[B]
+        val it = c.iterator
+        while (it.hasNext) buf += it.next()
+        buf
+    }
+  }
+
+  class ArrayBufferView[A](val elems: Array[AnyRef], val start: Int, val end: Int) extends RandomAccessView[A] {
+    def apply(n: Int) = elems(start + n).asInstanceOf[A]
+  }
+
+  case class StringView(s: String) extends RandomAccessView[Char] {
+    val start = 0
+    val end = s.length
+    def apply(n: Int) = s.charAt(n)
+  }
+
+  /** Concrete collection type: String */
+  implicit class StringOps(val s: String) extends AnyVal with Ops[Char] {
+    def iterator: Iterator[Char] = new StringView(s).iterator
+  }
+
+  implicit class StringMonoTransforms(val s: String)
+  extends AnyVal with MonoTransformsOfSeqs[Char, String] {
+    protected def coll: Iterable[Char] = StringView(s)
+    protected def fromIterable(it: Iterable[Char]): String = {
+      val sb = new StringBuilder
+      for (ch <- it) sb.append(ch)
+      sb.toString
+    }
+  }
+
+  implicit class StringPolyTransforms(val s: String)
+  extends AnyVal with PolyTransforms[Char, Seq] {
+    protected def coll = StringView(s)
+    protected def fromIterable[B](it: Iterable[B]): Seq[B] = List.fromIterable(it)
+    def map(f: Char => Char): String = {
+      val sb = new StringBuilder
+      for (ch <- s) sb.append(f(ch))
+      sb.toString
+    }
+    def flatMap(f: Char => String) = {
+      val sb = new StringBuilder
+      for (ch <- s) sb.append(f(ch))
+      sb.toString
+    }
+    def ++(xs: IterableOnce[Char]): String = {
+      val sb = new StringBuilder(s)
+      for (ch <- xs.iterator) sb.append(ch)
+      sb.toString
+    }
+    def ++(xs: String): String = s + xs
+  }
+
+  /* ------------ Views --------------------------------------- */
+
+  /** A lazy iterable */
+  trait View[+A] extends Iterable[A] with FromIterable[View] {
+    override def view = this
+    override def fromIterable[B](c: Iterable[B]) = c match {
+      case c: View[B] => c
+      case _ => View.fromIterator(c.iterator)
+    }
+  }
+
+  /** Iterator defined in terms of indexing a range */
+  trait RandomAccessView[+A] extends View[A] {
```

Comment:
AFAIK Rust initially had `RandomAccessIterator`, but then they decided to drop it. It was creating more problems than solving.